File: src/main/java/com/agentstore/common/dto/response/CommonResponse.java
package com.agentstore.common.dto.response;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.agentstore.common.exception.constants.ErrorCode;

@JsonInclude(JsonInclude.Include.ALWAYS)
public record CommonResponse<T>(
        boolean isSuccess,
        String message,
        String errorCode,
        T result
) {

    private static final String SUCCESS_MESSAGE = "요청이 성공했습니다.";

    public static CommonResponse<Void> emptySuccess() {
        return new CommonResponse<>(true, SUCCESS_MESSAGE, null, null);
    }

    public static <T> CommonResponse<T> success(T result) {
        return new CommonResponse<>(true, SUCCESS_MESSAGE, null, result);
    }

    public static CommonResponse<Void> failure(ErrorCode errorCode) {
        return failure(errorCode, errorCode.getMessage());
    }

    public static CommonResponse<Void> failure(ErrorCode errorCode, String message) {
        return new CommonResponse<>(false, message, errorCode.getCode(), null);
    }
}

File: src/main/kotlin/com/agentstore/common/dto/response/CommonResponse.kt
package com.agentstore.common.dto.response

import com.agentstore.common.exception.constants.ErrorCode
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty

@JsonInclude(JsonInclude.Include.ALWAYS)
data class CommonResponse<T>(
    @get:JsonProperty("isSuccess")
    val isSuccess: Boolean,
    val message: String,
    val errorCode: String?,
    val result: T?,
) {
    companion object {
        private const val SUCCESS_MESSAGE = "요청이 성공했습니다."

        fun emptySuccess(): CommonResponse<Void> {
            return CommonResponse(
                isSuccess = true,
                message = SUCCESS_MESSAGE,
                errorCode = null,
                result = null,
            )
        }

        fun <T> success(result: T): CommonResponse<T> {
            return CommonResponse(
                isSuccess = true,
                message = SUCCESS_MESSAGE,
                errorCode = null,
                result = result,
            )
        }

        fun failure(errorCode: ErrorCode): CommonResponse<Void> {
            return failure(errorCode = errorCode, message = errorCode.message)
        }

        fun failure(errorCode: ErrorCode, message: String): CommonResponse<Void> {
            return CommonResponse(
                isSuccess = false,
                message = message,
                errorCode = errorCode.code,
                result = null,
            )
        }
    }
}
